File: 0046-permutations/0046-permutations.py
from typing import List

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        def back(nums, l, p):
            if len(p) == len(nums):
                l.append(p[:])  # append a copy of current path
                return
            for i in range(len(nums)):
                if nums[i] in p:
                    continue
                p.append(nums[i])
                back(nums, l, p)
                p.pop()
        
        l = []  # not [[]]
        p = []
        back(nums, l, p)
        return l

File: 0046-permutations/0046-permutations.java
class Solution {
    public static void backtrack(List<List<Integer>> result,ArrayList<Integer> prem, int[] nums)
    {
        if(prem.size()==nums.length)
        {
            result.add(new ArrayList<>(prem));
            return;

        }

        for(int i=0;i<nums.length;i++)
        {
            if(prem.contains(nums[i]))
            continue;

            prem.add(nums[i]);

            backtrack(result, prem, nums);

            prem.remove(prem.size()-1);
        }
    }
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result=new ArrayList<>();
        backtrack(result, new ArrayList<>(), nums);
        return result;



    }
}